sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/streaming/StreamWriter.java

```diff
@@ -27,6 +28,9 @@
  *
  * Streaming queries are divided into intervals of data called epochs, with a monotonically
  * increasing numeric ID. This writer handles commits and aborts for each successive epoch.
+ *
+ * Note that StreamWriter implementations should provide instances of
+ * {@link StreamingDataWriterFactory}.
```

Comment:
What about adding `createStreamWriterFactory` that returns the streaming interface? That would make it easier for implementations and prevent throwing cast exceptions because a `StreamingDataWriterFactory` is expected.

Reply:
That wouldn't be compatible with SupportsWriteInternalRow. We could add a StreamingSupportsWriteInternalRow, but that seems much more confusing both for Spark developers and for data source implementers.

Reply:
What do you think about removing the `SupportsWriteInternalRow` and always using `InternalRow`? For the read side, I think using `Row` and `UnsafeRow` is a problem: https://issues.apache.org/jira/browse/SPARK-23325

I don't see the value of using `Row` instead of `InternalRow` for readers, so maybe we should just simplify on both the read and write paths.

Reply:
I'm broadly supportive. I'll detail my thoughts in the jira.